privacyidea/lib/policy.py

```diff
@@ -2441,6 +2447,29 @@ def user(cls, g, scope, action, user_object):
                    client=g.client_ip, action=action, adminrealm=None, time=None,
                    sort_by_priority=True)
 
+    @classmethod
+    def token(cls, g, scope, action, token_obj):
+        """
+        Match active policies with a scope, an action and a token object.
+        The client IP is matched implicitly.
+        From the token object we try to determine the user as the owner.
+        If the token has no owner, we try to determine the tokenrealm.
+        We fallback to realm=None
+        :param g: context object
+        :param scope: the policy scope. SCOPE.ADMIN cannot be passed, ``admin`` must be used instead.
+        :param action: the policy action
+        :param token_obj: The token where the user object or the realm should match.
+        :rtype: ``Match``
+        """
+        if token_obj.user:
+            return cls.user(g, scope, action, token_obj.user)
+        else:
+            realm = None
+            realms = token_obj.get_realms()
+            if len(realms) > 0:
+                realm = realms[0]
```

Comment:
This is a good question. I am currently not sure, in which kind of logic policy problem we might end up.
I will give it some more thoughts.

Reply:
My findings so far: ``Match.token`` is only used in ``generic_challenge_response_reset_pin``. So in 98% of the cases we will have a tokenowner. While validate/check can be called with a specific serial number it is the question, if it is really call with a serial number of a token, that has no owner.
It could however called from other places in the future.

We handle list of policies and contradicting policies at other places, but then only when certain attributes (like realm) is missing. I am not sure, if we should do a generic approach (allowing a list of realms in the underlying methods) or add the code here.

Adding the code here uglifies the code for a scenario, that is seldomly used.

Adding the code in the lower layers would add a flexibility and complexibility we have not missed till now.

Still need to give this some thinking. Maybe we should open another issue, if we might want to solve this generically in lower layers. What do you think?

Reply:
The realm in the policies is typically used synonymous to user realm and thus does not allow for multiple realms (We would need to change quite some code to support this).
On the other hand, we should not just silently use the first token realm (if there are several). At least we should log this somewhere just in case.

Reply:
You are right. I will check with other policies how realms are used.

We could: If token only has one tokenrealm, use this very realm.
If token has more than one tokenrealm, write a warning, or do exception or whatever.

Reply:
It looks like (not 100% investigated) we have no policy use with token realms. 
We are e.g. using ``Match.realm`` in the policy check for ACTION.MAXTOKENREALM. But in this case again it is not the *tokenrealm* but a *given* realm, to which the token is about to be enrolled.

So having a ``Match.token`` actually might not be a that good idea. Well, it is handy in these cases, where we *know*, that there is a user, but we only have the *token* of the user. It allows us to save code.

However, what would we do, if we have a call to ``Match.token`` if the token has no owner?

~~~~python
@classmethod
    def token(cls, g, scope, action, token_obj):
        if token_obj.user:
            return cls.user(g, scope, action, token_obj.user)
        else:
            WHAT TO DO NOW!?
~~~~

We could simply log a **warning** or raise a **ProgrammingError**?

@plettich What are your further thoughts?

Reply:
I addressed this in c3be0850

Reply:
I am ok with that.